src/vm/jithelpers.cpp

```diff
@@ -2215,66 +2141,87 @@ BOOL ObjIsInstanceOf(Object *pObject, TypeHandle toTypeHnd, BOOL throwCastExcept
-
-    // If we are trying to cast a proxy we need to delegate to remoting
-    // services which will determine whether the proxy and the type are compatible.
-    // Start by doing a quick static cast check to see if the type information captured in
-    // the metadata indicates that the cast is legal.
-    if (fromTypeHnd.CanCastTo(toTypeHnd))
+    if (pMT->IsArray())
     {
-        fCast = TRUE;
+        if (toTypeHnd.IsArray())
+        {
+            fCast = pMT->ArrayIsInstanceOf(toTypeHnd, /* pVisited */ NULL);
+        }
+        else
+        {
+            MethodTable* toMT = toTypeHnd.GetMethodTable();
+            if (toMT->IsInterface() && toMT->HasInstantiation())
+            {
+                fCast = pMT->ArraySupportsBizarreInterface(toMT, /* pVisited */ NULL);
+            }
+            else
+            {
+                fCast = pMT->CanCastToClassOrInterface(toMT, /* pVisited */ NULL);
+            }
+        }
     }
-    else
-#ifdef FEATURE_COMINTEROP
-    // If we are casting a COM object from interface then we need to do a check to see 
-    // if it implements the interface.
-    if (toTypeHnd.IsInterface() && fromTypeHnd.GetMethodTable()->IsComObjectType())
+    else if (toTypeHnd.IsTypeDesc())
     {
-        fCast = ComObject::SupportsInterface(obj, toTypeHnd.AsMethodTable());
+        CastCache::TryAddToCache(pMT, toTypeHnd, FALSE);
+        fCast = FALSE;
     }
-    else
-#endif // FEATURE_COMINTEROP
-    if (Nullable::IsNullableForType(toTypeHnd, obj->GetMethodTable()))
+    else if (pMT->CanCastToClassOrInterface(toTypeHnd.AsMethodTable(), /* pVisited */ NULL))
     {
-        // allow an object of type T to be cast to Nullable<T> (they have the same representation)
         fCast = TRUE;
     }
-#ifdef FEATURE_ICASTABLE
-    // If type implements ICastable interface we give it a chance to tell us if it can be casted 
-    // to a given type.
-    else if (toTypeHnd.IsInterface() && fromTypeHnd.GetMethodTable()->IsICastable())
+    else
     {
-        // Make actuall call to ICastableHelpers.IsInstanceOfInterface(obj, interfaceTypeObj, out exception)
-        OBJECTREF exception = NULL;
-        GCPROTECT_BEGIN(exception);
-        
-        PREPARE_NONVIRTUAL_CALLSITE(METHOD__ICASTABLEHELPERS__ISINSTANCEOF);
+#ifdef FEATURE_COMINTEROP
+        // If we are casting a COM object from interface then we need to do a check to see 
+        // if it implements the interface.
+        if (toTypeHnd.IsInterface() && pMT->IsComObjectType())
+        {
+            fCast = ComObject::SupportsInterface(obj, toTypeHnd.AsMethodTable());
+        }
+        else
+#endif // FEATURE_COMINTEROP
+            if (Nullable::IsNullableForType(toTypeHnd, obj->GetMethodTable()))
+            {
+                // allow an object of type T to be cast to Nullable<T> (they have the same representation)
+                CastCache::TryAddToCache(pMT, toTypeHnd, TRUE);
+                fCast = TRUE;
+            }
```

Comment:
With your change to methodtable.cpp to make the general purpose castclass logic handle Nullable, is this logic ever used?

Reply:
Yes, it looks like this can be removed.

---
In reply to: [306594685](https://github.com/dotnet/coreclr/pull/23548#discussion_r306594685) [](ancestors = 306594685)